core/builder.go

```diff
@@ -62,53 +86,71 @@ func defaultRepo(dstore ds.ThreadSafeDatastore) (repo.Repo, error) {
 	}, nil
 }
 
-func (nb *NodeBuilder) Online() *NodeBuilder {
-	nb.online = true
-	return nb
-}
+func NewNode(ctx context.Context, cfg *BuildCfg) (*IpfsNode, error) {
+	if cfg == nil {
+		cfg = new(BuildCfg)
+	}
 
-func (nb *NodeBuilder) Offline() *NodeBuilder {
-	nb.online = false
-	return nb
-}
+	err := cfg.fillDefaults()
+	if err != nil {
+		return nil, err
+	}
 
-func (nb *NodeBuilder) SetRouting(ro RoutingOption) *NodeBuilder {
-	nb.routing = ro
-	return nb
-}
+	n := &IpfsNode{
+		mode:      offlineMode,
+		Repo:      cfg.Repo,
+		ctx:       ctx,
+		Peerstore: peer.NewPeerstore(),
+	}
+	if cfg.Online {
+		n.mode = onlineMode
+	}
 
-func (nb *NodeBuilder) SetHost(ho HostOption) *NodeBuilder {
-	nb.peerhost = ho
-	return nb
-}
+	// TODO: this is a weird circular-ish dependency, rework it
+	n.proc = goprocessctx.WithContextAndTeardown(ctx, n.teardown)
 
-func (nb *NodeBuilder) SetRepo(r repo.Repo) *NodeBuilder {
-	nb.repo = r
-	return nb
-}
+	success := false
+	defer func() {
+		if !success {
+			n.teardown()
+		}
+	}()
 
-func (nb *NodeBuilder) NilRepo() *NodeBuilder {
-	nb.nilrepo = true
-	return nb
-}
+	// setup local peer ID (private key is loaded in online setup)
+	if err := n.loadID(); err != nil {
+		return nil, err
+	}
 
-func (nb *NodeBuilder) Build(ctx context.Context) (*IpfsNode, error) {
-	if nb.built {
-		return nil, ErrAlreadyBuilt
+	n.Blockstore, err = bstore.WriteCached(bstore.NewBlockstore(n.Repo.Datastore()), kSizeBlockstoreWriteCache)
+	if err != nil {
+		return nil, err
 	}
-	nb.built = true
-	if nb.repo == nil {
-		var d ds.Datastore
-		d = ds.NewMapDatastore()
-		if nb.nilrepo {
-			d = ds.NewNullDatastore()
-		}
-		r, err := defaultRepo(dsync.MutexWrap(d))
-		if err != nil {
+
+	if cfg.Online {
+		do := setupDiscoveryOption(n.Repo.Config().Discovery)
+		if err := n.startOnlineServices(ctx, cfg.Routing, cfg.Host, do); err != nil {
 			return nil, err
 		}
-		nb.repo = r
+	} else {
+		n.Exchange = offline.Exchange(n.Blockstore)
 	}
-	conf := standardWithRouting(nb.repo, nb.online, nb.routing, nb.peerhost)
-	return NewIPFSNode(ctx, conf)
+
+	n.Blocks = bserv.New(n.Blockstore, n.Exchange)
+	if err != nil {
+		return nil, err
+	}
+
+	n.DAG = dag.NewDAGService(n.Blocks)
+	n.Pinning, err = pin.LoadPinner(n.Repo.Datastore(), n.DAG)
+	if err != nil {
+		// TODO: we should move towards only running 'NewPinner' explicity on
+		// node init instead of implicitly here as a result of the pinner keys
+		// not being found in the datastore.
+		// this is kinda sketchy and could cause data loss
+		n.Pinning = pin.NewPinner(n.Repo.Datastore(), n.DAG)
+	}
+	n.Resolver = &path.Resolver{DAG: n.DAG}
+
+	success = true
```

Comment:
i c, I got this reversed. teardown is not called only for this path.

This can be used for `res.SetError`
```go
defer func(){
  if err != nil {
    res.SetError(err, cmds.ErrNormal)
  }
}()
```